src/storage/statistics_cl.c

```diff
@@ -508,3 +518,254 @@ stats_dump (const char *class_name_p, FILE * file_p)
+  att = (DB_ATTRIBUTE *) db_get_attributes_force (class_mop);
+  while (att != NULL)
+    {
+      class_attr_ndv->attr_cnt++;
+      att = db_attribute_next (att);
+    }
+
+  class_attr_ndv->attr_ndv = (ATTR_NDV *) malloc (sizeof (ATTR_NDV) * (class_attr_ndv->attr_cnt + 1));
+  if (class_attr_ndv->attr_ndv == NULL)
+    {
+      return ER_FAILED;
+    }
+
+  select_list = stats_make_select_list_for_ndv (class_mop, &class_attr_ndv->attr_ndv);
+  if (select_list == NULL)
+    {
+      return ER_FAILED;
+    }
+
+  /* create sampling SQL statement */
+  query = "SELECT /*+ SAMPLING_SCAN */ %s FROM [%s]";
+
+  buf_size = strlen (select_list) + strlen (class_name_p) + 40;
+  query_buf = (char *) malloc (sizeof (char) * buf_size);
+  if (query_buf == NULL)
+    {
+      return ER_FAILED;
+    }
+  snprintf (query_buf, buf_size, query, select_list, class_name_p);
+
+  if (file_p != NULL)
+    {
+      fprintf (file_p, "Query : %s\n", query_buf);
+    }
+
+  /* execute sampling SQL */
+  error = db_compile_and_execute_local (query_buf, &query_result, &query_error);
+  if (error < NO_ERROR)
+    {
+      goto end;
+    }
+
+  /* save result to NDV info */
+  error = db_query_first_tuple (query_result);
+  if (error != DB_CURSOR_SUCCESS)
+    {
+      goto end;
+    }
+
+  /* get NDV from tuple */
+  for (i = 0; i < class_attr_ndv->attr_cnt; i++)
+    {
+      error = db_query_get_tuple_value (query_result, i, &value);
+      if (error != NO_ERROR)
+	{
+	  goto end;
+	}
+      class_attr_ndv->attr_ndv[i].ndv = DB_GET_BIGINT (&value);
+    }
+
+  /* get count(*) */
+  error = db_query_get_tuple_value (query_result, i, &value);
+  if (error != NO_ERROR)
+    {
+      goto end;
+    }
+  class_attr_ndv->attr_ndv[i].ndv = DB_GET_BIGINT (&value);
+
+end:
+  if (select_list)
+    {
+      free_and_init (select_list);
+    }
+  if (query_buf)
+    {
+      free_and_init (query_buf);
+    }
+  if (query_result)
+    {
+      db_query_end (query_result);
+      query_result = NULL;
+    }
+
+  return error;
+}
+
+/*
+ * stats_make_select_list_for_ndv () - make select-list for ndv
+ *   return:
+ *   class_mop(in): ....
+ *   select_list(in/out):
+ */
+char *
+stats_make_select_list_for_ndv (const MOP class_mop, ATTR_NDV ** attr_ndv)
+{
+  DB_ATTRIBUTE *att;
+  char column[DB_MAX_IDENTIFIER_LENGTH + 20] = { '\0' };
+  DB_DOMAIN *dom;
+  size_t buf_size = 1024;
+  ATTR_NDV *att_ndv = *attr_ndv;
+  int i = 0;
+  char *select_list = NULL, *select_buf = NULL;
+
+  select_list = (char *) malloc (sizeof (char) * buf_size);
+  if (select_list == NULL)
+    {
+      return NULL;
+    }
+  /* init select_list */
+  *select_list = '\0';
+
+  /* make select_list */
+  att = (DB_ATTRIBUTE *) db_get_attributes_force (class_mop);
+  while (att != NULL)
+    {
+      /* check if type is varchar(4000) or lob. */
+      dom = db_attribute_domain (att);
+      if (TP_IS_LOB_TYPE (TP_DOMAIN_TYPE (dom)) || (TP_IS_CHAR_TYPE (TP_DOMAIN_TYPE (dom)) && dom->precision > 4000))
```

Comment:
A defined string instead of 4000 would be nice.

Reply:
I fixed it.